pkg/controller/petset/identity_mappers.go

```diff
@@ -0,0 +1,202 @@
+}
+
+type VolumeIdentityMapper struct {
+	ps *apps.PetSet
+}
+
+func (v *VolumeIdentityMapper) AddIdentity(id string, pet *api.Pod) {
+	petVolumes := []api.Volume{}
+	petClaims := GetClaims(id, v.ps)
+
+	// These volumes will all go down with the pod. If a name matches one of
+	// the claims in the pet set, it gets clobbered.
+	podVolumes := map[string]api.Volume{}
+	for _, podVol := range pet.Spec.Volumes {
+		podVolumes[podVol.Name] = podVol
+	}
+
+	// Insert claims for the idempotent petSet volumes
+	for name, claim := range petClaims {
+		// Volumes on a pet for which there are no associated claims on the
+		// petset are pod local, and die with the pod.
+		podVol, ok := podVolumes[name]
+		if ok {
+			glog.V(4).Infof("Overwriting existing volume source %v", podVol.Name)
+		}
+		newVol := api.Volume{
+			Name: name,
+			VolumeSource: api.VolumeSource{
+				PersistentVolumeClaim: &api.PersistentVolumeClaimVolumeSource{
+					ClaimName: claim.Name,
+					ReadOnly:  false,
+				},
+			},
+		}
+		petVolumes = append(petVolumes, newVol)
+	}
+
+	// Transfer any ephemeral pod volumes
+	for name, vol := range podVolumes {
+		if _, ok := petClaims[name]; !ok {
+			petVolumes = append(petVolumes, vol)
+		}
+	}
+	pet.Spec.Volumes = petVolumes
+	return
+}
+
+func (v *VolumeIdentityMapper) GetIdentity(pet *api.Pod) string {
+	// TODO: Make this a hash?
+	return v.String(pet)
+}
+
+func (v *VolumeIdentityMapper) String(pet *api.Pod) string {
+	ids := []string{}
+	petVols := sets.NewString()
+	for _, petVol := range v.ps.Spec.VolumeClaimTemplates {
+		petVols.Insert(petVol.Name)
+	}
+	for _, podVol := range pet.Spec.Volumes {
+		// Volumes on a pet for which there are no associated claims on the
+		// petset are pod local, and die with the pod.
+		if !petVols.Has(podVol.Name) {
+			continue
+		}
+		if podVol.VolumeSource.PersistentVolumeClaim == nil {
+			// TODO: Is this a part of the identity?
+			ids = append(ids, fmt.Sprintf("%v:None", podVol.Name))
+			continue
+		}
+		ids = append(ids, fmt.Sprintf("%v:%v", podVol.Name, podVol.VolumeSource.PersistentVolumeClaim.ClaimName))
+	}
+	sort.Strings(ids)
+	return strings.Join(ids, "")
+}
+
+func GetClaims(id string, ps *apps.PetSet) map[string]api.PersistentVolumeClaim {
+	petClaims := map[string]api.PersistentVolumeClaim{}
+	for _, pvc := range ps.Spec.VolumeClaimTemplates {
+		claim := pvc
+		// TODO: Name length checking
+		claim.Name = fmt.Sprintf("%v-%v-%v", claim.Name, ps.Name, id)
+		claim.Namespace = ps.Namespace
+		claim.Labels = ps.Spec.Selector.MatchLabels
+
+		// TODO: We're assuming that the claim template has a volume QoS key, eg:
+		// volume.alpha.kubernetes.io/storage-class: anything
+		petClaims[pvc.Name] = claim
+	}
+	return petClaims
+}
+
+func GetClaimsForPet(pet *api.Pod, ps *apps.PetSet) []api.PersistentVolumeClaim {
+	// Strip out the "-(index)" from the pet name and use it to generate
+	// claim names.
+	id := strings.Split(pet.Name, "-")
```

Comment:
Fair enough, it doesn't fit in the interface (identityMapper only gives/gets identity of single pets and their references, not claim objects) but it could well be a dangling method.